Mathlib/Computability/AkraBazzi.lean

```diff
@@ -0,0 +1,1836 @@
+      _ ≤ f u := by
+          rw [←Real.norm_of_nonneg (hg_pos u hu.1), ←Real.norm_of_nonneg (hf_pos u hu.1)]
+          exact h_lb u hu.1
+  case ub => calc
+    f u ≤ c₂ * g u := by
+                rw [←Real.norm_of_nonneg (hg_pos u hu.1), ←Real.norm_of_nonneg (hf_pos u hu.1)]
+                exact h_ub u hu.1
+      _ ≤ c₂ * (c₄ * g x) := by gcongr; exact (hg_bound u hu).2
+      _ = c₂ * c₄ * (c₁⁻¹ * c₁) * g x := by simp [c₁_cancel]; ring
+      _ = c₂ * c₄ * c₁⁻¹ * (c₁ * g x) := by ring
+      _ ≤ c₂ * c₄ * c₁⁻¹ * f x := by
+                gcongr
+                rw [←Real.norm_of_nonneg (hf_pos x hbx), ←Real.norm_of_nonneg (hg_pos x hbx)]
+                exact h_lb x hbx
+
+lemma of_isEquivalent {f g : ℝ → ℝ}
+    (hg : GrowsPolynomially g) (hf : f ~[atTop] g) (hf' : ∀ᶠ x in atTop, 0 ≤ f x)
+    (hg' : ∀ᶠ x in atTop, 0 ≤ g x) : GrowsPolynomially f :=
+  of_isTheta hg hf.isTheta hf' hg'
+
+lemma of_isEquivalent_const {f : ℝ → ℝ} {c : ℝ} (hc : 0 < c) (hf : f ~[atTop] fun _ => c) :
+    GrowsPolynomially f := by
+  refine of_isEquivalent GrowsPolynomially.const hf (hf.tendsto_const.eventually
+    (eventually_ge_nhds hc)) (eventually_of_forall (fun _ => ?_))
+  simp [le_of_lt hc]
+
+end GrowsPolynomially
+end AkraBazziRecurrence
+
+/-!
+#### Definition of Akra-Bazzi recurrences
+
+This section defines the predicate `AkraBazziRecurrence T g a b r` which states that `T`
+satisfies the recurrence
+T(n) = ∑_{i=0}^{k-1} a_i T(r_i(n)) + g(n)
+with appropriate conditions on the various parameters.
+-/
+
+/-- An Akra-Bazzi recurrence is a function that satisfies the recurrence
+`T n = (∑ i, a i * T (r i n)) + g n`. -/
+structure AkraBazziRecurrence {k : ℕ} (T : ℕ → ℝ) (g : ℝ → ℝ) (a : Fin k → ℝ)
+    (b : Fin k → ℝ) (r : Fin k → ℕ → ℕ) where
+  /-- Point below which the recurrence is in the base case -/
+  n₀ : ℕ
+  /-- n₀ is always > 0 -/
+  n₀_gt_zero : 0 < n₀
+  /-- There is at least one term -/
+  k_gt_zero : 0 < k
+  /-- The a's are nonzero -/
+  a_pos : ∀ i, 0 < a i
+  /-- The b's are nonzero -/
+  b_pos : ∀ i, 0 < b i
+  /-- The b's are less than 1 -/
+  b_lt_one : ∀ i, b i < 1
+  /-- `g` is nonnegative -/
+  g_nonneg : ∀ x ≥ 0, 0 ≤ g x
+  /-- `g` grows polynomially -/
+  g_grows_poly : AkraBazziRecurrence.GrowsPolynomially g
+  /-- The actual recurrence -/
+  h_rec (n : ℕ) (hn₀ : n₀ ≤ n) : T n = (∑ i, a i * T (r i n)) + g n
+  /-- Base case: T(n) > 0 whenever n < n₀ -/
+  T_gt_zero' (n : ℕ) (hn : n < n₀) : 0 < T n
+  /-- The r's always reduce n -/
+  r_lt_n : ∀ i n, n₀ ≤ n → r i n < n
+  /-- The r's approximate the b's -/
+  dist_r_b : ∀ i, (fun n => (r i n : ℝ) - b i * n) =o[atTop] fun n => n / (log n) ^ 2
+
+namespace AkraBazziRecurrence
+
+variable {k : ℕ} {T : ℕ → ℝ} {g : ℝ → ℝ} {a b : Fin k → ℝ} {r : Fin k → ℕ → ℕ}
+  (R : AkraBazziRecurrence T g a b r)
+
+/-- Smallest `b i` -/
+noncomputable def min_bi : Fin k :=
+  haveI : NeZero k := ⟨ne_of_gt R.k_gt_zero⟩
+  Classical.choose <| Finite.exists_min b
```

Comment:
This ought to be straightforward to obtain computably by using `Finset.univ.min` or `Nat.find`. I'm surprised we don't have `Fin.find` or similar.

Reply:
Ah, nevermind, this is a sequence of reals.